design/ceph/osd-migration.md

```diff
@@ -0,0 +1,187 @@
+
+**NOTE**: ceph-volume `activate` command doesn't require `OSD_STORE_FLAG` to be passed as an argument. It auto detects the backend store that was used during the OSD `prepare`
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a Ceph version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  store:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### Replace OSD
+Operator reconciler will replace one OSD at a time.
+A config map will be used to store the list of OSDs to be replaced and the OSD being replaced currently.
+OSD replacement steps:
+  1. List out all OSDs where `osd-store:<osd store type>` does not match with `spec.storage.store.type`.
+  2. If the PG status is not `active+clean`, don't replace any OSD.
+  3. If the PG status is `active+clean` but a previous OSD replacement is not completed, then don't replace any new OSD.
+  4. If the PG status is `active+clean` and no replacement is in progress, then select an OSD to be replaced from the list.
+  5. Delete the OSD deployment.
+  6. Create an OSD prepare job with an environment variable of OSD ID to be replaced.
+  7. OSD prepare pod will destroy the OSD and prepare it again using the same OSD ID. Refer `Destroy OSD` section below for details.
+  8. Once the destroyed OSD pod is added back, delete the configmap.
+  9. If there is any error during the OSD migration, then preserve the OSD being replaced in the configmap for next reconcile.
+  10. Reconcile the operator and perform as same steps until all the OSDs have migrated to the new backend store.
+
+#### Destroy OSD
+
+OSD prepare pod job will perform OSD deletion using following steps.
+- Check for OSD ID environment variable of the OSD to be destroyed.
+- Use `ceph volume list` to fetch the OSD path.
+- Destroy the OSD using following command:
+
+    ```
+    ceph osd destroy <OSD_ID> --yes-i-really-mean-it
+    ```
+
+- Wipe the OSD disk. This removes all the data on the device.
+- Prepare the OSD with the new store type by using the same OSD ID. This is done by passing the OSD ID as `--osd-id` flag to `ceph-volume` command.
+
+    ```
+    ceph-volume lvm prepare --osd-id <OSD_ID> --data <OSD_PATH>
+    ```
+
+#### Manage OSD migration status
+
+- OSD migration status will managed by `OSDReplaceConfig` API. This will be stored as the config map. For example:
+``` yaml
+apiVersion: v1
+data:
+  config: '{"osdList":{"osds":[{"id":1,"path":"/mnt/set1-data-1qwdpk","node":"minikube-m03"},{"id":2,"path":"/mnt/set1-data-2c8h9k","node":"minikube-m04"}]},"isPgHealthy":true,"OSDInprogress":{"id":0,"path":"/mnt/set1-data-045k6q","node":"minikube-m02"},"startTimeStamp":"2023-07-18T14:41:33.180909077Z"}'
```

Comment:
Is this example out of date? I didn't think the full list of OSDs was needed anymore in the configmap, only the osdInProgress. Similarly, line 90 above may be need updating if we're not storing the list of OSD IDs.

Reply:
yes it was out of date. Updated now.